tests/frontend/test_circuitspecs_x.py

```diff
@@ -0,0 +1,481 @@
+
+
+    def test_interferometers(self, tol):
+        """Test that the compilation correctly decomposes the interferometer using
+        the rectangular_symmetric mesh"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+
+        with prog.context as q:
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[0], q[4])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[1], q[5])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[2], q[6])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[3], q[7])
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | q
+
+        res = prog.compile("X")
+
+        expected = sf.Program(8)
+
+        with expected.context as q:
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[0], q[4])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[1], q[5])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[2], q[6])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[3], q[7])
+            ops.Interferometer(U, mesh="rectangular_symmetric", drop_identity=False) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U, mesh="rectangular_symmetric", drop_identity=False) | (q[4], q[5], q[6], q[7])
+            ops.MeasureFock() | q
+
+        expected = expected.compile(DummyCircuit())
+
+        assert program_equivalence(res, expected, atol=tol, compare_params=False)
+
+    def test_unitaries_do_not_match(self):
+        """Test exception raised if the unitary applied to modes [0, 1, 2, 3] is
+        different to the unitary applied to modes [4, 5, 6, 7]"""
+        prog = sf.Program(8)
+        U = random_interferometer(4)
+
+        with prog.context as q:
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[0], q[4])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[1], q[5])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[2], q[6])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[3], q[7])
+            ops.Interferometer(U) | (q[0], q[1], q[2], q[3])
+            ops.Interferometer(U) | (q[4], q[5], q[6], q[7])
+            ops.BSgate() | (q[2], q[3])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="The applied unitary on modes"):
+            res = prog.compile("X")
+
+    def test_unitary_too_large(self):
+        """Test exception raised if the unitary is applied to more
+        than just modes [0, 1, 2, 3] and [4, 5, 6, 7]."""
+        prog = sf.Program(8)
+        U = random_interferometer(8)
+
+        with prog.context as q:
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[0], q[4])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[1], q[5])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[2], q[6])
+            ops.S2gate(SQ_AMPLITUDE, 0) | (q[3], q[7])
+            ops.Interferometer(U) | q
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="The applied unitary cannot mix between the modes"):
+            res = prog.compile("X")
```

Comment:
Nice tests! However, perhaps there should also be some tests for `num_modes != 8` (e.g., `num_modes = 4`), just to make sure it works. There should also likely be a test that an exception is raised if the number of modes is not even

Reply:
All the ones that can maid to work for arbitrary number of modes now take an extra argument for this quantity.

Reply:
awesome